### PR TITLE
fix: harden strided-copy-based op kernels

### PR DESCRIPTION
**Pull request by rebel-clee**
## Summary of Changes

Hardens the three strided-copy-based op kernels (`cat`, `index_select`, `index_copy`) with an op-level CPU fallback and upstream-parity overlap checks. The fallback is enabled by default, gated on a new `strided_copy_error` category in `TORCH_RBLN_DISABLE_FALLBACK`.

The kernels have two robustness issues. First, the backend can fail on a large batched strided copy, breaking downstream model execution; the fallback keeps models running until the upstream backend is fixed. Second, the kernels lack the overlap checks that PyTorch's out variants enforce, so aliased inputs can produce silently incorrect results.

The fallback is narrowly scoped: only the batched-copy failure routes to CPU recomputation, while validation and other backend errors propagate so caller-side bugs still surface. The overlap checks exclude `index_copy`'s `self ↔ out` case, its legitimate in-place pattern.

---

## Type of Change

- [x] `core` — Changes to RBLN backend, device layer, C++ extension, op registration, or `torch.compile` integration
- [x] `fix` — Bug fix

---

## Affected Modules

- [x] Ops/Kernels
- [x] C++ extension (`torch_rbln/csrc`)
- [x] Docs

---

## Checklist

* [x] PR title follows [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) format
* [x] Linting passes — see [Linting](docs/LINTING.md)
* [x] All CI tests pass — see [CI/CD Workflows](docs/WORKFLOWS.md)
* [x] Relevant documentation has been updated (if applicable)

---

### CODE REVIEW

File: test/rbln/test_cat_index_select_v2v.py [hunk 1]
```diff
@@ -306,6 +315,30 @@ def test_cat_mixed_dtype_promotes_to_common(self):
...
+        out = torch.empty(4, 6, dtype=dtype, device=DEVICE).t()
+        self.assertFalse(out.is_contiguous())
+        torch.cat([to_dev(a_cpu), to_dev(b_cpu)], dim=0, out=out)
+        _check(out, expected)
+
+    @pytest.mark.usefixtures("enable_eager_malloc")
+    def test_large_strided_cat(self):
+        """``aten::cat`` over a large batched strided copy."""
+        # Baseline allocations exercise the bulk strided copy path.
+        baseline_allocs = [torch.empty(102400, 2560, dtype=torch.float16, device=DEVICE) for _ in range(2)]
```
Comment (rebel-chanheo) [inline test/rbln/test_cat_index_select_v2v.py:333]: How about adding a test case to test the environment variable strided_copy_error?

File: aten/src/ATen/native/rbln/RBLNStridedV2V.cpp [hunk 2]
```diff
@@ -108,7 +109,18 @@ void strided_v2v_copy(const at::Tensor& dst, const at::Tensor& src, c10::rbln::V
 void strided_v2v_copy(const at::Tensor& dst, const at::Tensor& src) {
   c10::rbln::V2VBatch batch;
   strided_v2v_copy(dst, src, batch);
-  batch.submit();
+  try {
+    batch.submit();
+  } catch (const std::exception& e) {
```
Comment (rebel-chanheo) [inline aten/src/ATen/native/rbln/RBLNStridedV2V.cpp]: Exception scope is too broad here.

catch (const std::exception&) swallows everything derived from std::exception not just the "batched dispatch too large" failure this is meant to handle. With the fallback default-on, all of these get silently rerouted to CPU.

Two problems:
- It undercuts the project's stance (TORCH_RBLN_DISABLE_FALLBACK=compile_error) that dispatch/compile failures should be loud, not silently masked.
- On a hot path like cat/index_select/index_copy, a silent CPU roundtrip is a hidden perf cliff — only a WARN log, throughput just drops.

Suggestion: narrow the catch to the specific error the backend raises for this case (type / error code / message match) and rethrow the rest.

File: c10/rbln/RBLNV2VBatch.cpp [hunk 3]
```diff
@@ -137,19 +137,29 @@ void V2VBatch::submit() {
...
+    } else {
+      // Heterogeneous — per-entry memcpy_v2v handles host-bounce internally.
+      RBLN_LOG_DEBUG("V2VBatch::submit draining {} entries (per-entry fallback)", impl_->pending.size());
+      for (const auto& e : impl_->pending) {
+        memcpy_v2v(e.dst, e.src, e.nbytes);
+      }
     }
+  } catch (...) {
+    reset_state();
+    throw;
```
Comment (rebel-chanheo) [inline c10/rbln/RBLNV2VBatch.cpp]: also, please check here

File: aten/src/ATen/native/rbln/RBLNTensorShape.cpp [hunk 4]
```diff
@@ -175,7 +176,24 @@ at::Tensor& cat_out_rbln(const at::ITensorListRef& tensors, int64_t dim, at::Ten
     strided_v2v_copy(dst_view, t, batch);
     axis_offset += extent;
   }
```
Comment (rebel-seungchul) [inline aten/src/ATen/native/rbln/RBLNTensorShape.cpp:178]: **Optional cleanup — duplicate submit/fallback block**

The `try { batch.submit(); } catch (...)` block here is repeated almost verbatim in `index_select_out_rbln` and `index_copy_out_rbln` (and a simpler variant in the 1-arg `strided_v2v_copy` overload). The only part that differs per call site is the CPU fallback body — the `is_fallback_disabled("strided_copy_error")` gate and the `RBLN_LOG_WARN` message are identical.

Consider factoring the common shell into a small helper in `RBLNStridedV2V.h` (all three TUs already include it), e.g.:

```cpp
template <typename Fn>
void submit_or_fallback(c10::rbln::V2VBatch& batch, const char* op_name, Fn&& cpu_fallback) {
  try {
    batch.submit();
  } catch (const std::exception& e) {
    if (c10::rbln::is_fallback_disabled("strided_copy_error")) {
      throw;
    }
    RBLN_LOG_WARN("{}: batched strided copy failed — falling back to CPU op. "
                  "Underlying error: {}", op_name, e.what());
    cpu_fallback();
  }
}
```

so each call site keeps only the op-specific logic:

```cpp
submit_or_fallback(batch, "cat_out_rbln", [&] {
  std::vector<at::Tensor> cpu_tensors;
  cpu_tensors.reserve(inputs.size());
  for (const at::Tensor& t : inputs) cpu_tensors.push_back(t.cpu());
  out.copy_(at::cat(cpu_tensors, axis));
});
```

This keeps the fallback policy (gate category + log wording) in one place, which helps if we later want to add a metric/counter on fallback. Non-blocking — feel free to skip if you'd rather keep the blocks inline.

(Note: the per-op overlap checks above are intentionally *not* worth factoring — they mirror upstream PyTorch's per-op `assert_no_overlap` argument sets 1:1, so spelling them out per kernel reads better.)

File: test/rbln/test_cat_index_select_v2v.py (hunk 1 above)
Comment (rebel-clee) [reply test/rbln/test_cat_index_select_v2v.py:333]: I omitted the test on purpose. Whether the intent is to verify the environment variable's routing on the happy path or to collect the cases that fail under it as expected failures, the same problem applies. Setting `TORCH_RBLN_DISABLE_FALLBACK=strided_copy_error` today triggers the rebel-compiler bug as an unhandled exception — exactly the diagnostic signal the environment variable produces. But a CI test that fails (or whose expected failure is the bug itself) goes stale the moment the fix lands.

When that fix lands, I plan to set `TORCH_RBLN_DISABLE_FALLBACK=strided_copy_error` as a default environment variable for the suite — same pattern as `TORCH_RBLN_DISABLE_FALLBACK=compile_error` — so the fallback path never engages silently in CI.

File: aten/src/ATen/native/rbln/RBLNStridedV2V.cpp (hunk 2 above)
Comment (rebel-clee) [reply aten/src/ATen/native/rbln/RBLNStridedV2V.cpp]: Thanks — agreed the broad catch swallows everything `RBLN_CHECK` raises, including caller-side validation errors that should propagate (nullptr inputs, dtype mismatches, etc.). I'd ideally narrow by type via a typed exception, but that touches the wrapper API beyond this PR's scope.

As a near-term workaround I've added a stable "rbln_memcpy_v2v_multi failed" error message to the wrapper's `RBLN_CHECK`, and the fallback now matches on that substring — validation errors no longer hit the CPU path.

File: c10/rbln/RBLNV2VBatch.cpp (hunk 3 above)
Comment (rebel-clee) [reply c10/rbln/RBLNV2VBatch.cpp]: Thanks for catching this. I've refactored `V2VBatch::submit()` to use an RAII guard for batch state reset — the guard's destructor now handles cleanup on both normal and exception paths, replacing the broad `catch (...)` and its catch-reset-rethrow wrapping.

File: aten/src/ATen/native/rbln/RBLNTensorShape.cpp (hunk 4 above)
Comment (rebel-clee) [reply aten/src/ATen/native/rbln/RBLNTensorShape.cpp:178]: Thanks — agreed the duplication is real. I've pulled `submit_or_fallback` into `RBLNStridedV2V.h` and `RBLNStridedV2V.cpp` along the lines you sketched. All three op kernels and the 1-arg overload now route through it.

**Review by rebel-seungchul — APPROVED** [review]
LGTM

**Review by rebel-chanheo — APPROVED** [review]
LGTM
